### PR TITLE
TypeScript Transition Start + Base Functionality

### PR DESCRIPTION
**Pull request by holgerd77**
This is a start on the TypeScript transition, PR currently depends on https://github.com/ethereumjs/ethereumjs-devp2p/pull/88 to be merged and released on the devp2p library. For continued work library has to be linked locally by going to the devp2p branch and doing "npm link" and then going to this client branch and doing "npm link ethereumjs-devp2p".

Goal of this first commit 0aac06899bc7986cb7764e3b7b461ecee8d98dba within the PR was to having both the build process for Node and browser as well as the CLI invocation of the client run without errors, all changes within the commit are triggered by errors which have been triggered along the way.

We might want to continue to work a bit on this on this branch, at least until the devp2p PR is released. Then maybe please drop a note that you start on something, try to enclose the part of the library you want to transition and publish as a PR against this branch.

Once devp2p PR is released we can very well merge here since @ryanio set this up for an evolved file-by-file transition and doing small PRs will likely be easier to review and integrate.

### CODE REVIEW

**Comment by holgerd77** [comment]
Posting here to avoid double work: will work on this during the next 2-3 hours.

**Comment by holgerd77** [comment]
Ok, I now also fixed the tests here which I somehow forgot on first run.

Some notes:

1. On a second thought it likely does not make sense to optimize / develop towards a `ts-node` invoked CLI client, so I revoked this. I have now developed so that the following commands work:

- `npm run build:node && node dist/bin/cli.js`
- `npm run build`
- `npm run test`

2. After an hour or so of chaos I've sticked to keep the exports as `CommonJS` exports with `module.exports` and not transition this on the first round while having mixed js and ts files since this always broke something (tests, browser, CLI) on some end. I would recommend that we generally keep these and do the module transition after we have everthing in TypeScript.

3. For the libp2p part (and eventually others, haven't checked?) there is heavy use of the [testdouble.js](https://github.com/testdouble/testdouble.js) package for creating mock objects to not use the real libp2p stack. This has gotten ES6 module support in [3.15.0](https://github.com/testdouble/testdouble.js/blob/master/CHANGELOG.md#3150) it is nevertheless tricky (not possible?) to get this running with `ts-node` since the node module loader has to be replaced (see PR from the release CHANGELOG), this also only works with Node 13+. For now I've commented out the respective tests, feel free to update with a better/working solution.

**Comment by holgerd77** [comment]
Tests are now passing again, issue is something in the CI setup.

**Comment by holgerd77** [comment]
@evertonfraga Could you have a short look at the CI setup here?

@ryanio Browser tests are not yet working, can you have a look at the Karma setup?

**Comment by evertonfraga** [comment]
@holgerd77 sure.

**Comment by evertonfraga** [comment]
After a first round of investigation, it seems that a failing `git cat-file` might be a github instability, So I'm re-running the tests.

**Comment by ryanio** [comment]
@holgerd77 @evertonfraga ah yes I think upgrading the `checkout` actions step to `v2` is the fix for that.

for the karma issue, this is in the ci logs:

`[framework.browserify]: Error: Can't walk dependency graph: Cannot find module '../../lib/blockchain' from '/home/runner/work/ethereumjs-client/ethereumjs-client/test/blockchain/chain.js'`

i guess it's having trouble importing a ts file from js.

ideally we can also convert the test suite to ts then i can update the karma config, right now it's a little complex by browserifying the js test suite (kind of how ethereumjs-vm was before we moved all over to typescript) so that would be a win for simplifying. otherwise we would have to point the tests to the `dist` files and requiring building before.

**Comment by holgerd77** [comment]
@ryanio Ah, yes, thanks, it was the `checkout` switch to `v2´, so CI for Node is now working, that's nice for some reassurance.

Then let's keep `Karma` out for now and also ignore the browser build and do the switch at the end of the transition I would say.

**Comment by holgerd77** [comment]
Just did some more transitions. Realizing that the scope of this is really pretty big, so I'll stick with a lot of `any`s for now, I think it's already be a substantial effort to get the base transition done and it will be likely something for a lot of further iterations to continue to evolve on typing.

One specifically useful thing would be to have types for the options at some point. I gave this a start but then realized quickly that this is also not yet practical at this point, since there is no such thing as a fixed set of options but these are individually mixed for each class. So this is likely something for some substantial follow-up refactoring.

**Comment by holgerd77** [comment]
Ok, will stop on this for today.

Some notes if someone wants to continue:

1. It turned out that TypeScript put everything under a global scope and I got a lot of "Cannot redeclare block-scoped variable" errors, basically on all library imports (like `BN`) within class files and most of generic constant definitions (like `defaultOptions`). A direct solution from [this article](https://medium.com/@muravitskiy.mail/cannot-redeclare-block-scoped-variable-varname-how-to-fix-b1c3d9cc8206) didn't work, but adopting this to assign classes to both `export` and `module.exports` (e.g. `export = module.exports = class Peer extends EventEmitter { }`) turned out to be practical and worked well.

2. if you stumble upon a "Default export of the module has or is using private name 'NodeEventTarget'" error along switching to such the export structure from above it is likely that there is a `import { EventEmitter } from 'events'` ES6-style import at the beginning of the file which needs to be converted back to the old CommonJS format `const { EventEmitter } = require('events')`.

**Comment by holgerd77** [comment]
Ok, this now includes the `ethereumjs-devp2p` `v3.0.3` dependency and is now installable/executable without any local linking.

**Comment by holgerd77** [comment]
Ok, code base is now transitioned except the `rpc` folder from the main source and the test files.

**Comment by cgewecke** [comment]
> the rpc folder from the main source and the test files

Posting to avoid duplication: will work on this this pm (PST)

**Comment by cgewecke** [comment]
Migrated the remaining files & tests to TS. 

Status: Tests pass but ... 
+ the coverage measurement has stopped working (no data is being collected)
+ there's an unhandled rejection warning btw test 212 & 213 (this predates my changes)

**Comment by holgerd77** [comment]
Great, already on the "good morning" side of things, hehe 😄, is it ok if I eventually pick this up or will you still continue to work on this?

**Comment by cgewecke** [comment]
@holgerd77 Yes! It's all yours! (I will check in the morning if there's anything left)

**Comment by ryanio** [comment]
nice :) fixing coverage should be easy, just need to add a .nycrc pointing to the source - I can do along with karma if nobody else does before then.

**Comment by holgerd77** [comment]
@ryanio yes, go ahead! 😀 Please let me know once you finish our stop with work.

**Comment by holgerd77** [comment]
I would like to work on transitioning some parts to ES6 imports and do deeper type integration, if you can push something early on which supports this (by tweaking the compiler options e.g.) that would be beneficial. I would generally start branching off (you can directly work on the branch) and then later open a PR towards this branch.

Let me generally know if this conflicts with your work plans though.

**Comment by ryanio** [comment]
ok you can go ahead right now, I am going to bed 😄 but I can pick up in my morning 👍

**Comment by holgerd77** [comment]
Ok, I got the Karma tests running by fixing an issue with a dynamic `require` in the code base, lot's of tests are still failing though.

I limited the test runs to one reference test to have something to start with, selected the `test/net/peerpool.js` file. The first error "Uncaught TypeError: Cannot read property \\'isBuffer\\' of undefined" preventing the tests to run can be attributed to the logger import. If the following lines get commented out this error goes away, didn't get to the root cause here though yet:

```typescript
const { defaultLogger } = require('../../lib/logging')
defaultLogger.silent = true
```

Execution goes then along to the first test and the error message then becomes:

"Error: Sorry, but CommonJS module replacement with td.replace() is only supported under Node.js runtimes."

This is actually all over the place - I already stumbled upon this before already and then commented out some libp2p tests as a temporary "solution". Think we have to dig deeper here though, since the [testdouble](https://github.com/testdouble/testdouble.js) library is pretty much deep rooted within the test setup. Related problem is using test double for replacing/mocking ES6 modules, see respective [docs](https://github.com/testdouble/testdouble.js/#tdreplace-and-tdreplaceesm-for-replacing-dependencies) section.

**Comment by holgerd77** [comment]
Some possible paths for the `testdouble` problem:

1. Only use selected unproblematic functionality from the library (so no `td.replace()` e.g.)
2. Evaluate and use an alternative library, main candidate: [Sinon.js](https://sinonjs.org/) (I assume?)
3. Use custom vanilla JS mock functionality and code

**Comment by cgewecke** [comment]
Posting to avoid duplication: will work on this this pm (PST)

**Comment by cgewecke** [comment]
@holgerd77 

Spent a bit of time looking a the browser tests from a different "single test" starting point: `test/blockchain/chain.ts`... 

There are errors coming from `ethereumjs-util` (of which there are 3 versions in the deps: v5, v6, v7) which look like this:
```
Uncaught AssertionError: The field transactionsTrie must have byte length of 32 FAILED
```

When I debugged this by going into node_modules to log the values I saw that karma-typescript is bundling with the top-level ethereum-util (v7) while the regular ts-node tests use a lower version, deeper in the dependency tree (local to the `blockchain` or `block` modules). 
 
TLDR I don't think karam-typescript is resolving sub-dependencies correctly.

Have begun migrating the module loading to TS in #145 on the theory that if imports are done normally, the recipes for karma and coverage which work at the VM and in `ethereumjs-util` will also work here.

**Comment by holgerd77** [comment]
@cgewecke ah ok, interesting. I was always unsure if we should first get the browser tests running again and then turn on to the ES6 transition or the otherway around and had a bit of a tendency for the first variant. Then I will also do some work on #145 (let me know if you are still working on it atm, currently assume that you are not), think we should be on the safe side and get things going again as long as the node tests are passing.

Have asked Vinay - who originally wrote most of this code here - on the testdouble question, here his answer:

>testdouble.js is not an absolute requirement for the tests but it saves a lot of effort in mocking out many of the dependencies.

>I believe the latest version of testdouble added support for ESM modules: 
https://github.com/testdouble/testdouble.js/blob/master/docs/7-replacing-dependencies.md#es-module-support

>The tests will have to be modified to use td.replaceEsm() instead of td.replace(). Another small annoyance is that replaceEsm is asynchronous whereas replace was synchronous, so will have to add some async/await wrappers.

>I think it will be easier to make these small changes than it will be to completely remove testdouble.

Might be that we can also keep selected functionality from testdouble and rather take this on more differentiated and look case-by-case what can be fixed and where the problems are. In general I am still a bit puzzled that testdouble was working also for the browser tests before the transition without having this "runs only in Node" behavior. Maybe it worth a test to pre-process the code also through browserify after karma-typescript and see what happens? 🤔

**Comment by holgerd77** [comment]
(@cgewecke but generally: very much not sure on this, totally don't want to withhold you from e.g. experimenting with an alternative library or the like)

**Comment by holgerd77** [comment]
Experimented a bit with `node -r ts-node/register --experimental-modules --loader=testdouble node_modules/tape/bin/tape 'test/!(integration)/**/*.+(js|ts)'` but this doesn't seem to be a stable and viable solution. Also generally seems to be a bit over the top to me to replace the whole module loader just for having a utility library working.

**Comment by cgewecke** [comment]
Not certain it should be done here, but another "change all files" preparatory PR might be to add prettier, lint the codebase, add lint check in CI, etc. 

@holgerd77 @ryanio WDYT

**Comment by ryanio** [comment]
@cgewecke I agree, I was just talking with Ev that it would be nice to release and integrate the config 2.0 packages into devp2p and client before bringing them into the monorepo.

**Comment by holgerd77** [comment]
@cgewecke yeah, that makes sense. Seems we have an agreement to merge here reading from the chat, so @ryanio or also @cgewecke maybe once you are are finished with the changes you are eventually currently still working on maybe just approve and merge here. Then we can do the linting in another PR.

**Comment by holgerd77** [comment]
@ryanio ah, maybe don't do force pushes here any more, that makes me a bit nervous 😜

**Comment by ryanio** [comment]
@holgerd77 oops, ok, noted.

I made some progress on the karma config, but now it's hanging in `chain.open()` => `chain.update()` => `chain.getLatestHeader()`.

Stepping into `getLatestHeader()` where it's getting stuck, first it seems that the blockchain init may not be fully working as expected since if I do

```console.log(this.blockchain._headHeader)``` => `null`

However if I do

```await promisify(this.blockchain._init.bind(this.blockchain))()``` then `this.blockchain._headHeader` populates.

Then I tried `this.blockchain.getBlock(this.blockchain._headHeader, console.log)` (which is what `getLatestHeader()` does) and I got the same error that @cgewecke encountered above:

`The field transactionsTrie must have byte length of 32`

**Review by ryanio — APPROVED** [review]
LGTM, exciting progress!

**Comment by cgewecke** [comment]
No further changes from me - and greenlight for merge as well.

**Review by evertonfraga — APPROVED** [review]
Can't wait to bring this codebase to the monorepo :)

**Comment by holgerd77** [comment]
Ok, have deactivated the browser check for now, will then merge here. 🎉
